### PR TITLE
active node inventory cleanup

### LINKED ISSUE #8834: Trim output of `ListNodes` to not show nodes that have hit TTL
DynamoDB can take up to 48 hours to TTL out a resource (like an SSH server) from the backend. 

From the [DynamoDB documentation](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/howitworks-ttl.html
):

```
Depending on the size and activity level of a table, the actual delete operation of an expired item
can vary. Because TTL is meant to be a background process, the nature of the capacity used to expire and
delete items via TTL is variable (but free of charge). TTL typically deletes expired items within 48
hours of expiration.
```

This means users can have stale nodes show up in the Web UI or `tsh ls` output. The default output for should be updated to trim the stale entries when presenting this to the user. Internally we should track the stale entries to ensure consistency in cache state.

### PR DESCRIPTION
**Pull request by fspmarshall**
This PR introduces custom in-cache expiration logic intended to help mitigate issues related to slow expiry of backend node resources.  Traditionally, we have tried to keep the caching layer as a "dumb" reflection of the event stream, and not attempt to guess what the real backend state should be.  This mirror-only model was important for keeping the replication model simple, and for ensuring that teleport didn't present inconsistent states in the event of performance issues (e.g. slow event propagation).

Unfortunately, the extremely lazy manner in which some backends may expire items (esp. dynamo) has become a significant issue in larger clusters (e.g. in high churn dynamo clusters, 100K+ expired node entries might be present, leading to very noticeable performance degradation).

This PR attempts to get the best of both worlds, expiring entries only if the real backend is lagging significantly behind in emitting expiry events, and only if said entries are outliers relative to other stale resources (helping us to avoid inconsistency in the event of a generally degraded event stream).

Fixes #8834 (though in a different manner than originally proposed).  We may still want to add the option to filter out offline nodes in the future, but it is less pressing now, and won't be needed until we decouple "offline" and "expiry" (something we do intend to do down the road).

Related to #10019 (further investigation required, but I believe we could reasonably extend this fix to cover that problem as well)

### CODE REVIEW

File: lib/cache/cache_test.go [hunk 1]
```diff
@@ -2178,6 +2178,53 @@ func TestDatabases(t *testing.T) {
...
+	clock.Advance(time.Minute * 50)
+	time.Sleep(checkInterval * 3)
+
+	nodes, err = p.cache.GetNodes(ctx, apidefaults.Namespace)
+	require.NoError(t, err)
+
+	// half the nodes are now "expired", but relative expiry waits until nodes are
+	// expired by an extra 10-20 minutes, so just require that we're in a roughly
+	// expected range.
+	require.True(t, len(nodes) < 100 && len(nodes) > 50, "node_count=%d", len(nodes))
```
Comment (Tener) [inline lib/cache/cache_test.go:2225]: Can we expand the test to check the node count at different points in time? For example:
- before relative expiry has a chance to kick in ("start")
- during the partial cleanup (with assertion as above) ("during") 
- far enough in the future that relative expiry removed all affected nodes ("end")

The test may be further expanded to observe the process at different times ("during-1", "during-2", ...) and show a relationship between those (the number of nodes is steadily decreasing).

These changes would make the expected behavior of the "relative expiry" feature explicit and tested.

File: lib/cache/cache.go [hunk 2]
```diff
@@ -952,6 +980,111 @@ func (c *Cache) fetchAndWatch(ctx context.Context, retry utils.Retry, timer *tim
...
+		// upstream auth cache to perform relative expiry and propagate the changes.
+		return 0, nil
+	}
+
+	// TODO(fspmarshall): Start using dynamic value once it is implemented.
+	gracePeriod := apidefaults.ServerAnnounceTTL
+
+	if c.target != "auth" {
+		return 0, nil
+	}
```
Comment (rosstimothy) [inline lib/cache/cache.go:1023]: nit: remove duplicate/old check
```suggestion
```

File: lib/cache/cache_test.go [hunk 3]
```diff
@@ -2178,6 +2178,53 @@ func TestDatabases(t *testing.T) {
...
+			t.Fatal("timeout waiting for event")
+		}
+	}
+
+	nodes, err := p.cache.GetNodes(ctx, apidefaults.Namespace)
+	require.NoError(t, err)
+	require.Len(t, nodes, 100)
+
+	clock.Advance(time.Minute * 50)
+	time.Sleep(checkInterval * 3)
```
Comment (rosstimothy) [inline lib/cache/cache_test.go:2217]: is there any way to avoid sleeping?

File: lib/cache/cache.go [hunk 4]
```diff
@@ -952,6 +980,111 @@ func (c *Cache) fetchAndWatch(ctx context.Context, retry utils.Retry, timer *tim
...
+		latestExp = now
+	}
+
+	var removed int
+	for _, node := range nodes {
+		if node.Expiry().IsZero() || node.Expiry().Add(gracePeriod).After(latestExp) {
+			continue
+		}
+
+		// if we get here, the node's expiration time was greater than `gracePeriod` earleir than our
```
Comment (rosstimothy) [inline lib/cache/cache.go:1065]:
```suggestion
		// if we get here, the node's expiration time was greater than `gracePeriod` and earlier than our
```

**Review by timothyb89 — APPROVED** [review]
No comments from me beyond the other reviews

File: lib/cache/cache.go [hunk 5]
```diff
@@ -901,6 +914,13 @@ func (c *Cache) fetchAndWatch(ctx context.Context, retry utils.Retry, timer *tim
 
 	retry.Reset()
 
+	relativeExpiryInterval := interval.New(interval.Config{
+		Duration:      c.Config.RelativeExpiryCheckInterval,
+		FirstDuration: utils.HalfJitter(c.Config.RelativeExpiryCheckInterval),
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/cache/cache.go:919]: Unfortunately, yes.  That would be the case even without jitter, and is somewhat true today anyway (poor auth server performance or backend API client throttling can lead to cache's being out of sync in time).

The best mitigation I could come up with was to have the auth server emit its expiry events to downstream caches rather than having downstream caches expire entries independently.  This limits the number of competing views considerably.  It also means that we'll be able to easily trace any bugs, since auth caches can be queried by `tctl`, but proxy caches cannot.

Disagreement could be further minimized by reducing the `RelativeCheckExpiryInterval` if need be.

File: lib/cache/cache.go [hunk 6]
```diff
@@ -952,6 +980,111 @@ func (c *Cache) fetchAndWatch(ctx context.Context, retry utils.Retry, timer *tim
...
+// performRelativeNodeExpiry performs a special kind of active expiration where we remove nodes
+// which are clearly stale relative to their more recently heartbeated counterparts as well as
+// the current time. This strategy lets us side-step issues of clock drift or general cache
+// staleness by only removing items which are stale from within the cache's own "frame of
+// reference".
+//
+// to better understand why we use this expiry strategy, its important to understand the two
+// distinct scenarios that we're trying to accommodate:
+//
+// 1. Expiry events are being emitted very lazily by the real backend (*hours* after the time
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/cache/cache.go:992]: We already filter expired items out of our dynamo queries.  The problem is with the event stream not keeping the cache up to date (since the expiry events can show up so late).

Active removal from the backend is a reasonable alternative, and a naive version of it was the first patch that was tried for the clusters that are currently experiencing issues due to this problem.  It resulted in very large spikes in backend usage and throttling errors.  While it is obviously possible to avoid this with the correct rate limiting, and by instructing users to up their backend limits, we're already seeing higher backend usage than we'd really like in large clusters.  Furthermore, concurrency is a factor in active removal from the backend, since auth servers would need to coordinate to limit concurrent deletion, and we'd need to have some protection against deletes racing with writes (not a super complex thing to fix, but it would require tinkering with our existing backend API as it doesn't really have a concept of conditional deletion at the moment).

Basically, I think this strategy is a safer bet for a time-sensitive patch release, since it doesn't affect shared state between auth servers, or run the risk of triggering backend throttling in clusters with tight margins.

**Review by rosstimothy — APPROVED** [review]
